{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import python packages\n",
    "import tangram as tg\n",
    "from anndata import AnnData\n",
    "import sys, os\n",
    "sys.path.insert(0, \"FISHscale\")\n",
    "from FISHscale.utils import dataset\n",
    "import scanpy as sc\n",
    "sc.settings.verbosity = 0\n",
    "sc.set_figure_params(dpi = 450, dpi_save = 450)\n",
    "import numpy as np\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.spatial import KDTree\n",
    "from sklearn.preprocessing import scale\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import to_rgb, hex2color\n",
    "import pandas as pd\n",
    "\n",
    "# Set plotting settings\n",
    "import matplotlib as mpl\n",
    "mpl.rcParams['figure.facecolor'] = \"none\"\n",
    "mpl.rcParams['legend.labelcolor'] = \"black\"\n",
    "mpl.rcParams['text.color'] = \"black\"\n",
    "\n",
    "if not os.path.exists('figures'):\n",
    "   os.makedirs('figures')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read spatial data\n",
    "d = dataset.Dataset(\"Data/LBEXP20211113_EEL_HE_5w_970um_RNA_transformed_assigned.parquet\",\n",
    "                       gene_label = 'decoded_genes', \n",
    "                       x_label = 'r_transformed', \n",
    "                       y_label = 'c_transformed', \n",
    "                       pixel_size='0.27 micrometer', \n",
    "                       other_columns=['Brain'],\n",
    "                       z = 970,\n",
    "                       reparse=False)\n",
    "d.set_working_selection(\"Brain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating AnnData object from fishscale spatial data\n",
    "\n",
    "# Make hexagonal bins, the output can be used to create adata object\n",
    "# Returns pandas count dataframe to df_hex and spatial coordinates (x, y) to a list variable\n",
    "df_hex, coordinates = d.hexbin_make(spacing=50, min_count=10)\n",
    "\n",
    "# Initialize variables\n",
    "coordinates_to_keep = []\n",
    "count_df = df_hex.T\n",
    "spots_to_remove = []\n",
    "x, y = [], []\n",
    "\n",
    "# The spatial data contains part (hindbrain) which doesn't occur in our mistr data \n",
    "# Using the spatial coordinates (x, y) we are able to select the spots we are interested in \n",
    "for i, spot in enumerate(count_df.index):\n",
    "    if coordinates[i][0] < 1800 and coordinates[i][1] < 4300:\n",
    "        spots_to_remove.append(spot)\n",
    "    else:\n",
    "        coordinates_to_keep.append([coordinates[i][0]*-1,coordinates[i][1]])\n",
    "        x.append(coordinates[i][0]*-1)\n",
    "        y.append(coordinates[i][1])\n",
    "\n",
    "# Remove the unnecessary rows from df based on the spots_to_remove index list\n",
    "count_df = count_df.drop(spots_to_remove)\n",
    "\n",
    "# Initialize anndata with count data, and add the spatial cordinates \n",
    "adata_spatial = AnnData(count_df.to_numpy(), obsm={\"spatial\": np.array(coordinates_to_keep)})\n",
    "adata_spatial.var.index = np.array(count_df.columns)\n",
    "adata_spatial.obs.index = np.array(count_df.index)\n",
    "adata_spatial.obs[\"x\"] = x\n",
    "adata_spatial.obs[\"y\"] = y\n",
    "\n",
    "# Plot results after removing the parts not occuring in our single cell data\n",
    "with plt.rc_context({\"figure.figsize\": [3.5, 2.2],   \"figure.dpi\": 400}):\n",
    "    xs = adata_spatial.obs.x.values\n",
    "    ys = adata_spatial.obs.y.values\n",
    "    plt.axis(\"off\")\n",
    "    plt.scatter(xs, ys, s=0.8, color=\"#D3D3D3\");\n",
    "    plt.savefig(\"figures/raw_spatial_data.png\", dpi=450, bbox_inches='tight')\n",
    "    plt.savefig(\"figures/raw_spatial_data.svg\", dpi=450, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tangram plotting function inverts the y axis, and there is no parameter to control this\n",
    "# Therefore this function can be used to invert the y axis before passing the data to tangram plotting functions\n",
    "# and when tangram inverts the axis again, the plot is displayed correctly\n",
    "\n",
    "def invert_y_axis_for_tangram(adata):\n",
    "    counter, inverted_coordinates = 0, []\n",
    "    # Looping over the coordinatas, and multiplying the y-axis coordinates by 1 and append the updated coordinates to list\n",
    "    for i in range(len(adata.obsm[\"spatial\"])):\n",
    "        inverted_coordinates.append((adata.obsm[\"spatial\"][i][0],adata.obsm[\"spatial\"][i][1]*-1))\n",
    "\n",
    "    # Store the updated coordinates in adata_spatial object\n",
    "    adata.obsm[\"spatial\"] = np.array(inverted_coordinates)\n",
    "    return adata\n",
    "\n",
    "# Calling the function that invers y axis, both with the full and subsetted spatial data\n",
    "adata_spatial = invert_y_axis_for_tangram(adata_spatial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code is edited version of tangram plot_genes_sc function\n",
    "# The original function contained some shortcommings,\n",
    "# and therefore a customized version of it was used to plot mesured / projected genes on space\n",
    "# The original function allows plotting measured and projected expression side by side,\n",
    "# whereas this function plots either measured or projected genes\n",
    "# Other changes: possibility to save the figure and some additional parameters that serves better our purposes\n",
    "\n",
    "from scipy.sparse.csr import csr_matrix\n",
    "from scipy.sparse.csc import csc_matrix\n",
    "from matplotlib.gridspec import GridSpec\n",
    "import pylab as pl\n",
    "\n",
    "sc.set_figure_params(dpi_save = 500)\n",
    "sc.set_figure_params(scanpy=True, fontsize=10)\n",
    "\n",
    "def construct_obs_plot(df_plot, adata, perc=0, suffix=None):\n",
    "    # clip\n",
    "    df_plot = df_plot.clip(df_plot.quantile(perc), df_plot.quantile(1 - perc), axis=1)\n",
    "\n",
    "    # normalize\n",
    "    df_plot = (df_plot - df_plot.min()) / (df_plot.max() - df_plot.min())\n",
    "\n",
    "    if suffix:\n",
    "        df_plot = df_plot.add_suffix(\" ({})\".format(suffix))\n",
    "    adata.obs = pd.concat([adata.obs, df_plot], axis=1)\n",
    "\n",
    "def convert_adata_array(adata):\n",
    "    if isinstance(adata.X, csc_matrix) or isinstance(adata.X, csr_matrix):\n",
    "        adata.X = adata.X.toarray()\n",
    "        \n",
    "def tangram_plot_genes_custom(adata, genes=[], x=\"x\",y = \"y\",spot_size=None, save=False,scale_factor=None, ncols=4, cmap=\"inferno\",\n",
    "                              perc=0,alpha_img=1.0,bw=False,return_figure=False, plot_measured_spatial=False, show=True):\n",
    "    # construct df_plot\n",
    "    data = []\n",
    "    \n",
    "    # remove df_plot in obs\n",
    "    if plot_measured_spatial == False:\n",
    "        adata.obs.drop([\"{} (projected gene expression)\".format(gene) for gene in genes], inplace=True, errors=\"ignore\", axis=1)\n",
    "        adata.var.index = [g.upper() for g in adata.var.index]\n",
    "        adata.obs.drop([\"{} (projected gene expression)\".format(gene) for gene in genes], inplace=True, errors=\"ignore\", axis=1)\n",
    "        \n",
    "        df = pd.DataFrame(data=np.array(adata[:, genes].X), columns=genes, index=adata.obs.index)\n",
    "        construct_obs_plot(df, adata, perc=perc, suffix=\"projected gene expression\")\n",
    "            \n",
    "    else:\n",
    "        adata.obs.drop([\"{} (measured gene expression)\".format(gene) for gene in genes], inplace=True, errors=\"ignore\", axis=1)\n",
    "        \n",
    "        # prepare adatas\n",
    "        convert_adata_array(adata)\n",
    "        adata.var.index = [g.upper() for g in adata.var.index]\n",
    "        adata.obs.drop([\"{} (measured gene expression)\".format(gene) for gene in genes],inplace=True,errors=\"ignore\",axis=1,)\n",
    "        for ix, gene in enumerate(genes):\n",
    "            if gene not in adata.var.index:\n",
    "                data.append(np.zeros_like(np.array(adata[:, 0].X).flatten()))\n",
    "            else:\n",
    "                data.append(np.array(adata[:, gene].X).flatten())\n",
    "\n",
    "        df = pd.DataFrame(\n",
    "            data=np.array(data).T, columns=genes, index=adata.obs.index,\n",
    "        )\n",
    "        construct_obs_plot(df, adata, suffix=\"measured gene expression\")\n",
    "\n",
    "    fig = plt.figure(figsize=(7, len(genes) * 3.5))\n",
    "    gs = GridSpec(len(genes), 2, figure=fig)\n",
    "    \n",
    "    #non visium data\n",
    "    if 'spatial' not in adata.obsm.keys():\n",
    "        #add spatial coordinates to obsm of spatial data \n",
    "        if plot_measured_spatial == True:\n",
    "            coords = [[x,y] for x,y in zip(adata.obs[x].values,adata.obs[y].values)]\n",
    "            adata.obsm['spatial'] = np.array(coords)\n",
    "        else:\n",
    "            coords = [[x,y] for x,y in zip(adata.obs[x].values,adata.obs[y].values)]\n",
    "            adata.obsm['spatial'] = np.array(coords)\n",
    "\n",
    "    if (\"spatial\" not in adata.uns.keys()) and (spot_size==None and scale_factor==None):\n",
    "        raise ValueError(\"Spot Size and Scale Factor cannot be None when ad_sp.uns['spatial'] does not exist\")\n",
    "        \n",
    "    gene_list=[]\n",
    "    for gene in genes:\n",
    "        if plot_measured_spatial==True:       \n",
    "            gene_list.append(\"{} (measured gene expression)\".format(gene))          \n",
    "        else:\n",
    "            gene_list.append(\"{} (projected gene expression)\".format(gene))\n",
    "        \n",
    "    if plot_measured_spatial == True:\n",
    "        sc.pl.spatial(adata, spot_size=spot_size, scale_factor=scale_factor, color=gene_list,\n",
    "            frameon=False, show=show, cmap=cmap, alpha_img=alpha_img, bw=bw, ncols=ncols, colorbar_loc=None, save=save)\n",
    "    else: \n",
    "        #key = dict_keys[i+j]\n",
    "        #sc.pl.spatial(adata, spot_size=spot_size, scale_factor=scale_factor, color=[\"{} (predicted)\".format(gene)],\n",
    "        sc.pl.spatial(adata, spot_size=spot_size, scale_factor=scale_factor, color=gene_list, \n",
    "            frameon=False, show=show, cmap=cmap, alpha_img=alpha_img, bw=bw, ncols=ncols, colorbar_loc=None,save=save )\n",
    "    \n",
    "    if show == True:\n",
    "        a = np.array([[0,1]])\n",
    "        pl.figure(figsize=(0.14, 4.5))\n",
    "        img = pl.imshow(a, cmap=\"YlGnBu\")\n",
    "        pl.gca().set_visible(False)\n",
    "        cax = pl.axes([0.1, 0.2, 0.8, 0.6])\n",
    "        pl.colorbar(orientation=\"vertical\", cax=cax)\n",
    "        pl.savefig(\"figures/YlGnBu_colorbar_1.png\", dpi=450, bbox_inches='tight')\n",
    "\n",
    "        a = np.array([[0,1]])\n",
    "        pl.figure(figsize=(0.14, 1.8))\n",
    "        img = pl.imshow(a, cmap=\"YlGnBu\")\n",
    "        pl.gca().set_visible(False)\n",
    "        cax = pl.axes([0.1, 0.2, 0.8, 0.6])\n",
    "        pl.colorbar(orientation=\"vertical\", cax=cax)\n",
    "        pl.savefig(\"figures/YlGnBu_colorbar_2.png\", dpi=450, bbox_inches='tight')\n",
    "    \n",
    "    # remove df_plot in obs\n",
    "    if plot_measured_spatial == True:\n",
    "        adata.obs.drop([\"{} (measured gene expression)\".format(gene) for gene in genes], inplace=True, errors=\"ignore\", axis=1)\n",
    "    else:\n",
    "        adata.obs.drop([\"{} (projected gene expression)\".format(gene) for gene in genes], inplace=True, errors=\"ignore\", axis=1)\n",
    "          \n",
    "    if return_figure==True:\n",
    "        return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the gene expression in spatial data\n",
    "for gene in ['OTX2', 'WNT1', 'FGF8', 'FGF17', 'FGF18']:  \n",
    "    tangram_plot_genes_custom(adata_spatial, genes=[gene],spot_size=50,\n",
    "                          scale_factor=0.1, perc = 0.001, return_figure=False, cmap=\"YlGnBu\", plot_measured_spatial=True, save=\"_measured_png/_\"+gene+\"_measured_expression.png\", show=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-bonefight_env]",
   "language": "python",
   "name": "conda-env-.conda-bonefight_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
